docs/content/docs/concepts/glossary.md

```diff
@@ -25,182 +25,605 @@ under the License.
 
 # Glossary
 
+#### Aggregation
+
+Aggregation is an operation that takes multiple values and returns a single value. When working with 
+streams, it generally makes more sense to think in terms of aggregations over finite windows, rather 
+than over the entire stream.
+
+#### (Flink) Application
+
+A Flink application is any user program that submits one or multiple [Flink Jobs](#flink-job) from its
+`main()` method. The execution of these jobs can happen in a local JVM or on a remote setup of clusters 
+with multiple machines.
+
+The jobs of an application can either be submitted to a long-running [Session Cluster](#session-cluster),
+to a dedicated [Application Cluster](#application-cluster), or to a [Job Cluster](#job-cluster).
+
+#### Application Cluster
+
+A Flink application cluster is a dedicated [Flink cluster](#(flink)-cluster) that only executes 
+[Flink jobs](#flink-job) from one [Flink application](#(flink)-application). The lifetime of the Flink
+cluster is bound to the lifetime of the Flink application.
+
+#### Asynchronous Snapshotting
+
+A form of [snapshotting](#snapshot) that doesn't impede the ongoing stream processing by allowing an 
+operator to continue processing while it stores its state snapshot, effectively letting the state 
+snapshots happen asynchronously in the background.
+
+#### At-least-once
+
+A fault-tolerance guarantee and data delivery approach where multiple attempts are made at delivering
+an event such that at least one succeeds. This guarantees that nothing is lost, but you may experience 
+duplicated results.
+
+#### At-most-once
+
+A data delivery approach where each event is delivered zero or one times. There is lower latency but
+events may be lost.
+
+#### Backpressure
+
+A situation where a system is receiving data at a higher rate than it can process during a temporary 
+load spike.
+
+#### Barrier Alignment
+
+For providing exactly-once guarantees, Flink aligns the streams at operators that receive multiple 
+input streams, so that the snapshot will reflect the state resulting from consuming events from both 
+input streams up to (but not past) both barriers. 
+
+#### Batch Processing
+
+This is the processing and analysis on a set of data that have already been stored over a period 
+of time (i.e. in groups or batches). The results are usually not available in real-time. Flink 
+executes batch programs as a special case of streaming programs.
+
+#### Bounded Streams
+
+Bounded [DataStreams](#datastream) have a defined start and end. They can be processed by ingesting 
+all data before performing any computations. Ordered ingestion is not required to process bounded streams 
+because a bounded data set can always be sorted. Processing of bounded streams is also known as 
+[batch processing](#batch-processing).
```

Comment:
I think for a glossary item this is actually pretty OK. We can't explain everything in detail in a glossary and I don't really know how to make this one better.

Reply:
Then I rather prefer to remove everything here that talks about the optimization, rather than giving this false impression that Flink will always optimize these streams.

Maybe a short but more correct wording could simply be:

> When doing stream processing, Flink can apply optimizations to bounded streams, while when doing batch processing, Flink requires that every input stream is a bounded stream.